### ISSUE TITLE
Problem with Priority Log

### ISSUE THREAD

**Issue by And-Garcia** [fonol/anki-search-inside-add-card#93]
Regards. I'm testing out the add-on for the first time. It works correctly on a clean install. I created a couple of test notes (text with some html, haven't done PDFs yet) and everything went fine with the queue system. However, after restarting Anki, I receive the following error:

> Debug info:
> Anki 2.1.22 (0ecc189a) Python 3.8.0 Qt 5.14.1 PyQt 5.14.1
> Platform: Windows 10
> Flags: frz=True ao=True sv=1
> Add-ons, last update check: 2020-05-09 05:59:42
> 
> Caught exception:
> Traceback (most recent call last):
>   File "aqt\progress.py", line 76, in handler
>   File "aqt\main.py", line 158, in setupProfileAfterWebviewsLoaded
>   File "aqt\main.py", line 195, in setupProfile
>   File "<C:\Program Files\Anki\decorator.pyc:decorator-gen-38>", line 2, in loadProfile
>   File "lib\site-packages\anki\hooks.py", line 666, in decorator_wrapper
>   File "lib\site-packages\anki\hooks.py", line 661, in repl
>   File "aqt\main.py", line 370, in loadProfile
>   File "aqt\gui_hooks.py", line 1275, in __call__
>   File "C:\Users\AG\AppData\Roaming\Anki2\addons21\1781298089\src\notes.py", line 301, in recalculate_priority_queue
>     current = _get_priority_list_with_last_prios()
>   File "C:\Users\AG\AppData\Roaming\Anki2\addons21\1781298089\src\notes.py", line 769, in _get_priority_list_with_last_prios
>     res = conn.execute(sql).fetchall()
> sqlite3.OperationalError: no such table: queue_prio_log

Some kind of issue with the priority log. Anki remains usable after this error and the search functionalities of the add-on continue to work as intended; however, notes are nowhere to be found. Additionally, attempts to access the Queue no longer prompt the original 'Queue is empty' message. I have attempted to delete search-data.db but the issue remains unchanged. Only way I've succeeded in restoring the app to work is by removing and re-installing it. Still, it only works correctly for one sitting and the issue re-occurs on Anki's next restart.

I'm uncertain of how to proceed at this point. Any help would be appreciated.

I have SQLite FTS5, thought I'd mention it since it's not in the Debug info.

**Comment by And-Garcia** [comment]
Messing with my install and back-ups, I managed to pinpoint that the issue actually stemmed from my configuration. I started from defaults and adjusted options one by one to determine the exact issue. Turns out the problem was related to the `Addon Note DB Folder Path` option. I had selected a custom path to store my notes and the add-on was correctly generating a `siac-notes.db` there. However, I failed to consider that the notes I had generated in the very same session where I set-up the path would not be properly saved to my custom folder. Thus, upon start-up, the file on my DB folder was being checked for notes that it knew nothing about. This is embarrassing because I had been changing the DB path on every clean install without thinking. Simply copying the DB file from user_files into my custom folder was enough to resolve the issue.

Apologies for any inconvenience caused.

**Comment by fonol** [comment]
Thanks for the detailed report. Happy that you could resolve it, but maybe it points to a deeper issue. At startup, it should check for the database tables to be there, and if not, create them. So the error you got there should not be possible, even considering your explanation in the second post. I will look into it, thanks anyways!

**Comment by fonol** [comment]
I just tested, I changed the folder path in the settings dialog, then restarted. It seemed to work for me.

I think in general, I should state more clearly on that config option that it does not copy your existing data in the new folder. So if you already have a siac-notes.db file, you'd have to manually move it to the new location.

**Comment by And-Garcia** [comment]
I'm pretty sure in my last test I did just that - change the path then restart, from a clean install. Then again, I did so much copying and pasting of the `siac-notes.db` that I wouldn't be surprised if I had forgotten to delete the db file from the custom folder.

I did a new test just now, to confirm. Started again from a clean install of the add-on. Upon first installation only `data.json` can be found in user_files. After first start, `siac-notes.db` is generated in user_files and `search-data.db` follows shortly after. I changed the DB folder from settings, and upon restart, a new `siac-notes.db` is generated in the previously especified folder path, as expected. However, the error does still pop-up. I also noted that the new db file is different in size from the original. Might be because the add-on is not getting a chance to actually build the tables on the former? I still don't exactly see why that would be the case...

If it works for you, it may simply be a quirk with my set-up. Normal behavior returns, like before, when I move the db file from user_files manually to my custom folder path. I've otherwise not ran into any problems and am very much enjoying the add-on so far, so there's that.

Thanks a lot for looking into it, either way!

**Comment by fonol** [comment]
That is so weird, because the code executed on start of the add-on literally is:

` 
    #create the file if not existing, check if all tables are there, if not, create them
    create_db_file_if_not_exists()
    ... two more lines
    # this is where your error happens
    gui_hooks.profile_did_open.append(recalculate_priority_queue)`

Anyways, a new version is up, I have only slight hopes, but you could try it nevertheless :)

**Comment by And-Garcia** [comment]
The behavior remains as I described despite the update.

I tried a couple more things. 

First, I tested changing the DB file path directly on `meta.json` after installation but before the first run. In this scenario, the generation of the DB file on my custom folder went correctly and it persisted after restart. I tried, however, then removing the custom folder option and restarting - a new DB file was generated in user_files, smaller in size, and the table not found error reared its head again. Roles reversed. It appears the problem is explicitly with generating tables after the first iteration. The first time is always fine, but attempts to rebuild can't go through.  

What's more, I found that Anki did not allow update or removal of the add-on in this state, that is, it appears having a `siac-notes.db` in user_files, whether it is being used or not, is essential for Anki to recognize the structure of the add-on. it might be a good idea to generate a place-holder there when not found, even if the DB folder path option is filled in.

To confirm the 'only on first iteration' thought, I did a new install with default settings, deleted the DB file from user_files and found the same characteristic behavior upon attempting to generate a new one - that is, the table not found error.

I had a hunch so I tested on a separate install of Anki with version 2.1.26. It worked fine even after deleting and regenerating `siac-notes.db`. Thus, it appears that the error is born my usage of version 2.1.22. I gather that this incompatibility is due to the database changes that ocurred in 2.1.24. You had already mentioned it [https://github.com/fonol/anki-search-inside-add-card/issues/91](here) so I'm sure you were already aware. Guess we didn't catch it here due to the puzzling way in which it manifested?

Either way, it appears this is entirely my fault for not documenting myself on the version changes!
